File: composer/modules/web/src/plugins/ballerina/model/tree/abstract-tree/select-clause-node.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import _ from 'lodash';
import Node from '../node';

class AbstractSelectClauseNode extends Node {


    setSelectExpressions(newValue, silent, title) {
        const oldValue = this.selectExpressions;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.selectExpressions = newValue;

        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'selectExpressions',
                    newValue,
                    oldValue,
                },
            });
        }
    }

    getSelectExpressions() {
        return this.selectExpressions;
    }


    addSelectExpressions(node, i = -1, silent) {
        node.parent = this;
        let index = i;
        if (i === -1) {
            this.selectExpressions.push(node);
            index = this.selectExpressions.length;
        } else {
            this.selectExpressions.splice(i, 0, node);
        }
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'child-added',
                title: `Add ${node.kind}`,
                data: {
                    node,
                    index,
                },
            });
        }
    }

    removeSelectExpressions(node, silent) {
        const index = this.getIndexOfSelectExpressions(node);
        this.removeSelectExpressionsByIndex(index, silent);
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'child-removed',
                title: `Removed ${node.kind}`,
                data: {
                    node,
                    index,
                },
            });
        }
    }

    removeSelectExpressionsByIndex(index, silent) {
        this.selectExpressions.splice(index, 1);
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'child-removed',
                title: `Removed ${this.kind}`,
                data: {
                    node: this,
                    index,
                },
            });
        }
    }

    replaceSelectExpressions(oldChild, newChild, silent) {
        const index = this.getIndexOfSelectExpressions(oldChild);
        this.selectExpressions[index] = newChild;
        newChild.parent = this;
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'child-added',
                title: `Change ${this.kind}`,
                data: {
                    node: this,
                    index,
                },
            });
        }
    }

    replaceSelectExpressionsByIndex(index, newChild, silent) {
        this.selectExpressions[index] = newChild;
        newChild.parent = this;
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'child-added',
                title: `Change ${this.kind}`,
                data: {
                    node: this,
                    index,
                },
            });
        }
    }

    getIndexOfSelectExpressions(child) {
        return _.findIndex(this.selectExpressions, ['id', child.id]);
    }

    filterSelectExpressions(predicateFunction) {
        return _.filter(this.selectExpressions, predicateFunction);
    }


    setGroupBy(newValue, silent, title) {
        const oldValue = this.groupBy;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.groupBy = newValue;

        this.groupBy.parent = this;

        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'groupBy',
                    newValue,
                    oldValue,
                },
            });
        }
    }

    getGroupBy() {
        return this.groupBy;
    }



    setHaving(newValue, silent, title) {
        const oldValue = this.having;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.having = newValue;

        this.having.parent = this;

        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'having',
                    newValue,
                    oldValue,
                },
            });
        }
    }

    getHaving() {
        return this.having;
    }




    isSelectAll() {
        return this.selectAll;
    }

    setSelectAll(newValue, silent, title) {
        const oldValue = this.selectAll;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.selectAll = newValue;
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'selectAll',
                    newValue,
                    oldValue,
                },
            });
        }
    }

}

export default AbstractSelectClauseNode;

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/SelectClauseNode.java
/*
 *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 *  WSO2 Inc. licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except
 *  in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.ballerina.compiler.syntax.tree;

import io.ballerina.compiler.internal.parser.tree.STNode;

import java.util.Objects;

/**
 * This is a generated syntax tree node.
 *
 * @since 2.0.0
 */
public class SelectClauseNode extends ClauseNode {

    public SelectClauseNode(STNode internalNode, int position, NonTerminalNode parent) {
        super(internalNode, position, parent);
    }

    public Token selectKeyword() {
        return childInBucket(0);
    }

    public ExpressionNode expression() {
        return childInBucket(1);
    }

    @Override
    public void accept(NodeVisitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <T> T apply(NodeTransformer<T> visitor) {
        return visitor.transform(this);
    }

    @Override
    protected String[] childNames() {
        return new String[]{
                "selectKeyword",
                "expression"};
    }

    public SelectClauseNode modify(
            Token selectKeyword,
            ExpressionNode expression) {
        if (checkForReferenceEquality(
                selectKeyword,
                expression)) {
            return this;
        }

        return NodeFactory.createSelectClauseNode(
                selectKeyword,
                expression);
    }

    public SelectClauseNodeModifier modify() {
        return new SelectClauseNodeModifier(this);
    }

    /**
     * This is a generated tree node modifier utility.
     *
     * @since 2.0.0
     */
    public static class SelectClauseNodeModifier {
        private final SelectClauseNode oldNode;
        private Token selectKeyword;
        private ExpressionNode expression;

        public SelectClauseNodeModifier(SelectClauseNode oldNode) {
            this.oldNode = oldNode;
            this.selectKeyword = oldNode.selectKeyword();
            this.expression = oldNode.expression();
        }

        public SelectClauseNodeModifier withSelectKeyword(
                Token selectKeyword) {
            Objects.requireNonNull(selectKeyword, "selectKeyword must not be null");
            this.selectKeyword = selectKeyword;
            return this;
        }

        public SelectClauseNodeModifier withExpression(
                ExpressionNode expression) {
            Objects.requireNonNull(expression, "expression must not be null");
            this.expression = expression;
            return this;
        }

        public SelectClauseNode apply() {
            return oldNode.modify(
                    selectKeyword,
                    expression);
        }
    }
}
